node/network/availability-recovery/src/lib.rs

```diff
@@ -0,0 +1,756 @@
+
+	/// The root of the erasure encoding of the para block.
+	erasure_root: Hash,
+
+	/// The chunks that we have received from peers.
+	received_chunks: HashMap<PeerId, ErasureChunk>,
+
+	/// The chunk requests that are waiting to complete.
+	requesting_chunks: FuturesUnordered<Timeout<oneshot::Receiver<ChunkResponse>>>,
+}
+
+impl Interaction {
+	async fn run(mut self) {
+		loop {
+			// If it's empty and requesting_chunks is empty,
+			// break and issue a FromInteraction::Concluded(RecoveryError::Unavailable)
+			if self.requesting_chunks.is_empty() && self.shuffling.is_empty() {
+				if self.to_state.send(FromInteraction::Concluded(
+					self.candidate_hash,
+					Err(RecoveryError::Unavailable),
+				)).await.is_err() {
+					return;
+				};
+				break;
+			}
+
+			while self.requesting_chunks.len() < N_PARALLEL {
+				if let Some(validator_index) = self.shuffling.pop() {
+					let (tx, rx) = oneshot::channel();
+
+					if self.to_state.send(FromInteraction::MakeRequest(
+						self.validators[validator_index as usize].clone(),
+						self.candidate_hash.clone(),
+						validator_index,
+						tx,
+					)).await.is_err() {
+						return;
+					};
+
+					self.requesting_chunks.push(rx.timeout(CHUNK_REQUEST_TIMEOUT));
+				} else {
+					break;
+				}
+			}
+
+			// Check if the requesting chunks is not empty not to poll to completion.
+			if !self.requesting_chunks.is_empty() {
+				// Poll for new updates from requesting_chunks.
+				while let Some(request_result) = self.requesting_chunks.next().await {
+					match request_result {
+						Some(Ok(Ok((peer_id, chunk)))) => {
+							// Check merkle proofs of any received chunks, and any failures should
+							// lead to issuance of a FromInteraction::ReportPeer message.
+							if let Ok(anticipated_hash) = branch_hash(
+								&self.erasure_root,
+								&chunk.proof,
+								chunk.index as usize,
+							) {
+								let erasure_chunk_hash = BlakeTwo256::hash(&chunk.chunk);
+
+								if erasure_chunk_hash != anticipated_hash {
+									if self.to_state.send(FromInteraction::ReportPeer(
+											peer_id,
+											COST_MERKLE_PROOF_INVALID,
+									)).await.is_err() {
+										return;
+									}
+								} else {
+									self.received_chunks.insert(peer_id, chunk);
+								}
+							} else {
+								if self.to_state.send(FromInteraction::ReportPeer(
+										peer_id,
+										COST_MERKLE_PROOF_INVALID,
+								)).await.is_err() {
+									return;
+								}
+							}
+						}
+						Some(Err(e)) => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								err = ?e,
+								"A response channel was cacelled while waiting for a chunk",
+							);
+						}
+						Some(Ok(Err(e))) => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								err = ?e,
+								"A chunk request ended with an error",
+							);
+						}
+						None => {
+							tracing::debug!(
+								target: LOG_TARGET,
+								"A chunk request has timed out",
```

Comment:
We need some way of triggering cleanup of the state which holds the other side of the channel, as mentioned above. A periodic poll of the state could work, because `oneshot::Sender` has an `is_cancelled` function.